### PR TITLE
Adding custom edge weight cost function capability

### PR DESCRIPTION
**Pull request by BrettHemes-3M**
First pass at providing users the ability to specify custom cost functions for planning graph edge weights based on the robot's joint space configuration.

### CODE REVIEW

**Comment by shaun-edwards** [comment]
@Jmeyer1292, please review.  @phvass FYI.

**Comment by phvass** [comment]
Glad to see this enhancement to Descartes. Thanks @BrettHemes !

**Comment by Jmeyer1292** [comment]
@BrettHemes Thanks for the PR. I'm all in favour of this, and my testing shows no measurable performance or functional regression of any sort when using your custom cost functions.

Given that, I'm going to go through and nitpick on style (even if much of the older code gave it no heed). I appreciate your expertise and your patience.

File: descartes_planner/include/descartes_planner/dense_planner.h [hunk 1]
```diff
@@ -22,6 +22,8 @@ class DensePlanner: public descartes_core::PathPlannerBase
   virtual ~DensePlanner();
 
   virtual bool initialize(descartes_core::RobotModelConstPtr model);
+  virtual bool initialize(descartes_core::RobotModelConstPtr model,
+                          descartes_planner::CostFunction cost_fucntion_callback);
```
Comment (Jmeyer1292) [inline descartes_planner/include/descartes_planner/dense_planner.h]: `cost_fucntion_callback` -> `cost_function_callback`. Type in `fucntion`.

File: descartes_planner/src/planning_graph.cpp [hunk 2]
```diff
@@ -998,15 +1005,20 @@ PlanningGraph::LinearWeightResult PlanningGraph::linearWeight(const JointTraject
...
     }
 
-    double vector_diff = 0;
-    for (unsigned i = 0; i < start_vector.size(); i++)
+    if (custom_cost_function_)
     {
-      double joint_diff = std::abs(end_vector[i] - start_vector[i]);
-      vector_diff += joint_diff ;
+      result.second = custom_cost_function_(start_vector, end_vector);
+    }else{
```
Comment (Jmeyer1292) [inline descartes_planner/src/planning_graph.cpp]: While you're at it, please put the braces on their own lines, e.g.

```
}
else
{
```

Just to keep with ROS style (I've been slowly trying to fix these issues in the older Descartes code).

**Comment by Jmeyer1292** [comment]
@BrettHemes If you have time, please fix the above small issues & I will merge right away.

**Comment by BrettHemes-3M** [comment]
@Jmeyer1292, style issues addressed.

**Comment by Jmeyer1292** [comment]
Thanks

**Comment by Jmeyer1292** [comment]
@BrettHemes Thank you for contribution.

**Pull request merged**